lib/verdict/storage/cookie_storage.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+module Verdict
+  module Storage
+    # CookieStorage, unlike other Verdict storage classes, is distributed and stored on the client.
+    # Because cookies are opaque to users, we obsfucate information about the test such as the
+    # human readable names for the experiment or assignment group. This means this class assumes
+    # that `name` will be an obsfucated value, or one that we comfortable being "public".
+    class CookieStorage < BaseStorage
+      DEFAULT_COOKIE_LIFESPAN_SECONDS = 15778476 # 6 months
+
+      attr_accessor :cookies
+      attr_reader :cookie_lifespan
+
+      def initialize(cookies: {}, cookie_lifespan: DEFAULT_COOKIE_LIFESPAN_SECONDS)
+        @cookies = cookies
+        @cookie_lifespan = cookie_lifespan
+      end
+
+      def store_assignment(assignment)
+        hash = { group: assignment.group.name, created_at: assignment.created_at.strftime('%FT%TZ') }
+        set("verdict_#{assignment.experiment.name}", nil, JSON.dump(hash))
+      end
+
+      def retrieve_assignment(experiment, subject)
+        if (value = get("verdict_#{experiment.name}", nil))
+          hash = parse_cookie_value(value)
+          group = experiment.groups.values.find { |g| g.name == hash['group'] }
+
+          if group.nil?
+            experiment.remove_subject_assignment(subject)
+            return nil
+          end
+
+          experiment.subject_assignment(subject, group, Time.xmlschema(hash['created_at']))
+        end
+      end
+
+      def remove_assignment(experiment, _subject)
+        remove("verdict_#{experiment.name}", nil)
+      end
+
+      def retrieve_start_timestamp(_experiment)
+        nil
+      end
+
+      def store_start_timestamp(_experiment, _timestamp)
+        true
+      end
+
+      protected
+
+      def get(scope, _key)
```

Comment:
The `_key` is here as parameter because it's present in the parent class?

Even if it's not needed right now for the cookie implementation, I am a bit worried this could cause issues later on, because this breaks the Liskov substitution principle.

Reply:
> The `_key` is here as parameter because it's present in the parent class?

Yeah. I match the interface of all methods in the `BaseStorage` class in `CookieStorage`.

> breaks the Liskov substitution principle.

Could you describe further, I think I'm missing something 😅I thought keeping interface identical helped maintain Liskov because `Experiment` can call any `Storage` instance with the same parameters and not get errors. Or is this because these are protected methods and `Experiment` shouldn't be hitting them directly anyway?

Reply:
https://github.com/Shopify/verdict/blob/f69077b7db8fcac0587b62b4873c63591011de1d/lib/verdict/storage/redis_storage.rb#L13-L17

Looking at the redis implementation we need to allow setting of multiple keys for a given scope. Or, maybe this is another distributed datastore difference and we don't care about that they key?

Reply:
> Or, maybe this is another distributed datastore difference and we don't care about that they key?

At the moment, the way it's implemented in Brochure is we have one cookie per assignment. I guess technically we could have a single cookie for verdict and have multiple assignments in that one cookie. 

I'm indifferent, happy to implement either way :)

Reply:
I think we use multiple keys per subject to be able to store additional metadata for an experiment, e.g. https://github.com/Shopify/verdict/blob/e32195bd397436c701430c2d200cbd09a2a24b00/lib/verdict/storage/base_storage.rb#L43-L46. 

As you already mentioned, this doesn't really work for cookie storage because it's not central storage.

However, it people would end up using this class expecting that to work, it will now silently fail, which violates the principle of the abstraction. Can we raise an exception in case the cookie storage object is used in a non-compatible way? Or maybe, should we change the storage abstraction?

Reply:
I've done two things:

- Refactored `CookieStorage` to use scope/key params (arbitrary, but it's less snowflakey)
- Now raises `NotImplementedError` error if someone tries to set the timestamp